### PR TITLE
Fix #8584: Duck spawn search now uses entire map

### LINKED ISSUE #8584: Ducks are not spawning!
**OS:** Windows 10
**Version:** 0.2.1
**Commit/Build:** 746feaa

Ducks no longer seem to spawn!

**Steps to reproduce:**
1. Create large body of water (code says you need at least 25 tiles of water at same height in a 7x7 area)
2. Wait forever
3. No ducks seem to spawn

### PR DESCRIPTION
**Pull request by ZehMatt**
This one should be without crashes. I left a few asserts in there to be sure and they have no performance impact.

### CODE REVIEW

File: src/openrct2/scenario/Scenario.cpp [hunk 1]
```diff
@@ -467,7 +472,13 @@ static int32_t scenario_create_ducks()
         r = scenario_rand();
         x2 = (r >> 16) & 0x7F;
         y2 = (r & 0xFFFF) & 0x7F;
-        create_duck(x + x2 - 64, y + y2 - 64);
+
+        int32_t targetX = x + x2 - 64;
```
Comment (duncanspumpkin) [inline src/openrct2/scenario/Scenario.cpp]: Create a CoordsXY for this.

File: src/openrct2/scenario/Scenario.cpp [hunk 2]
```diff
@@ -423,11 +424,10 @@ static int32_t scenario_create_ducks()
 {
     int32_t i, j, r, c, x, y, waterZ, centreWaterZ, x2, y2;
 
-    r = scenario_rand();
-    x = ((r >> 16) & 0xFFFF) & 0x7F;
-    y = (r & 0xFFFF) & 0x7F;
-    x = (x + 64) * 32;
-    y = (y + 64) * 32;
+    x = 64 + (scenario_rand_max(MAXIMUM_MAP_SIZE_PRACTICAL) * 32);
+    y = 64 + (scenario_rand_max(MAXIMUM_MAP_SIZE_PRACTICAL) * 32);
```
Comment (duncanspumpkin) [inline src/openrct2/scenario/Scenario.cpp]: CoordsXY this

**Comment by IntelOrca** [comment]
If this gets rebased and changed to use coords which might be easier now after @duncanspumpkin's refactor, we can merge.

**Comment by ZehMatt** [comment]
Sorry got a little distracted by other things, will do this today.

**Comment by ZehMatt** [comment]
Applied the requested changes and cleaned up the code a little bit. Also I bumped up the network version since scenario_rand is now called twice.

File: src/openrct2/scenario/Scenario.cpp [hunk 3]
```diff
@@ -415,60 +416,66 @@ void scenario_update()
...
             if (waterZ == centreWaterZ)
-                c++;
+                waterTiles++;
 
-            x2 += 32;
+            innerPos.x += 32;
         }
-        x2 -= 224;
-        y2 += 32;
+        innerPos.x -= 224;
```
Comment (duncanspumpkin) [inline src/openrct2/scenario/Scenario.cpp]: 7*32

**Comment by Gymnasiast** [comment]
Could you rebase this and rebump the network version? I think it's time to get this merged, and Duncan's requested changes have been made anyway.

**Comment by ZehMatt** [comment]
Rebased and incremented the network version

File: src/openrct2/scenario/Scenario.cpp [hunk 4]
```diff
@@ -415,60 +416,69 @@ void scenario_update()
...
-
 /**
  *
  *  rct2: 0x006744A9
  */
 static int32_t scenario_create_ducks()
 {
-    int32_t i, j, r, c, x, y, waterZ, centreWaterZ, x2, y2;
+    CoordsXY centrePos;
+    centrePos.x = 64 + (scenario_rand_max(MAXIMUM_MAP_SIZE_PRACTICAL) * 32);
```
Comment (Gymnasiast) [inline src/openrct2/scenario/Scenario.cpp]: It looks like this shifts the area from being in the centre and skipping the outer ring of invisible tiles, to one tile in both x and y directions. AFAICS, this should be changed to `32 + ...`.

Comment (ZehMatt) [reply src/openrct2/scenario/Scenario.cpp]: You have to consider that it subtracts 64 some lines below so that would go negative

**Pull request merged**